firewood/src/hashednode.rs

```diff
@@ -312,104 +311,140 @@ impl HasUpdate for Vec<u8> {
+        let key_bit_len = BITS_PER_NIBBLE * self.key.clone().count() as u64;
+        add_varint_to_buf(buf, key_bit_len);
 
         // Add key to hash pre-image
-        while let Some(high_nibble) = key_nibbles_iter.next() {
-            let low_nibble = key_nibbles_iter.next().unwrap_or(0);
+        while let Some(high_nibble) = self.key.next() {
+            let low_nibble = self.key.next().unwrap_or(0);
             let byte = (high_nibble << 4) | low_nibble;
-            hasher.update([byte]);
+            buf.update([byte]);
         }
     }
 }
 
-fn add_value_to_hasher<H: HasUpdate, T: AsRef<[u8]>>(hasher: &mut H, value: Option<T>) {
-    let Some(value) = value else {
-        let value_exists: u8 = 0;
-        hasher.update([value_exists]);
-        return;
+/// Returns the SHA-256 hash of the `preimage`.
+pub(crate) fn _hash<K: Iterator<Item = u8> + Clone, V: AsRef<[u8]>>(
+    preimage: HashPreimage<'_, K, V>,
+) -> TrieHash {
+    let mut hasher: Sha256 = Sha256::new();
+    preimage.write(&mut hasher);
+    hasher.finalize().into()
+}
+
+/// Writes the pre-image of `node`, which is at `path_prefix`, to `buf`.
+fn write_hash_preimage<H: HasUpdate>(node: &Node, path_prefix: &Path, buf: &mut H) {
+    let key = path_prefix
+        .iter()
+        .chain(node.partial_path().iter())
+        .copied();
+
+    let children = match *node {
+        Node::Branch(ref branch) => Some(&branch.child_hashes),
+        Node::Leaf(_) => None,
     };
 
-    let value_exists: u8 = 1;
-    hasher.update([value_exists]);
+    let value = match *node {
+        Node::Branch(ref branch) => branch.value.as_ref().map(|v| v.as_ref()),
+        Node::Leaf(ref leaf) => Some(leaf.value.as_ref()),
+    };
 
-    let value = value.as_ref();
+    match value {
+        Some(value) if value.len() >= 32 => {
+            let value_hash = Sha256::digest(value);
+            HashPreimage {
+                key,
+                value_digest: Some(value_hash),
+                children,
+            }
+            .write(buf);
+        }
+        _ => {
+            HashPreimage {
+                key,
+                value_digest: value,
+                children,
+            }
+            .write(buf);
+        }
+    }
+}
 
-    if value.len() >= 32 {
-        let value_hash = Sha256::digest(value);
-        add_len_and_value_to_hasher(hasher, &value_hash);
-    } else {
-        add_len_and_value_to_hasher(hasher, value);
+/// Returns the hash of `node` which is at `path_prefix`.
+pub fn hash_node(node: &Node, path_prefix: &Path) -> TrieHash {
+    let mut hasher: Sha256 = Sha256::new();
+    write_hash_preimage(node, path_prefix, &mut hasher);
+    hasher.finalize().into()
+}
+
+/// Returns the serialized representation of `node` used as the pre-image
+/// when hashing the node. The node is at the given `path_prefix`.
+pub fn hash_preimage(node: &Node, path_prefix: &Path) -> Box<[u8]> {
+    let mut buf = vec![];
```

Comment:
nit: Maybe `vec::with_capacity(node.value.len())` or something to avoid reallocating the buffer every time.

Reply:
Changed to pre-allocate capacity